DO NOT MERGE [Durable Objects] Make Durable Objects a tile (#9477)
cloudflare/cloudflare-docs: content/durable-objects/get-started.md

Reviewer: We should align this with the recently updated D1 tutorial at https://developers.cloudflare.com/d1/get-started/ - e.g.

1. using `create cloudflare` to set up a project.
2. moving any account / subscription requirements into a "Prerequisites" section
3. breaking into it more steps

6. Received an HTTP response back to the client with `return response`.

Refer to [Access a Durable Object from a Worker](/durable-objects/how-to/access-durable-object-from-a-worker/) to learn more about communicating to a Durable Object.

## 5. Configure Durable Object bindings

[Bindings](/workers/configuration/bindings/) allow your Workers to interact with resources on the Cloudflare developer platform. Configure Durable Object bindings in your Worker project's `wrangler.toml` by providing a binding name (for this guide, use `EXAMPLE_CLASS`) and the class name (`DurableObjectExample`).

```toml
---
filename: wrangler.toml
---
[[durable_objects.bindings]]
name = "EXAMPLE_CLASS"
class_name = "DurableObjectExample"
```

The `[[durable_objects.bindings]]` section contains the following fields:

  - `name` - Required. The binding name to use within your Worker.
  - `class_name` - Required. The class name you wish to bind to.

>   - `script_name` - Optional. Defaults to the current [environment's](/workers/wrangler/environments/) Worker script.

Reviewer 2: This should link to /durable-objects/platform/environments/.
```suggestion
  - `script_name` - Optional. Defaults to the current [environment's](/durable-objects/platform/environments/) Worker script.
```

 1. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com/).
 2. Go to **Workers & Pages** > **Plans**. 
 3. Select **Purchase Workers Paid** and complete the payment process to enable Durable Objects.

 ## 2. Create a Worker project

You will access your Durable Object from a [Worker](/workers/).

To create a Worker project, run:

{{<tabs labels="npm | yarn">}}
{{<tab label="npm" default="true">}}

```sh
$ npm create cloudflare@latest
```

{{</tab>}}
{{<tab label="yarn">}}

```sh
$ yarn create cloudflare@latest
```

{{</tab>}}
{{</tabs>}}

In your terminal, you will be asked a series of questions related to your project: 

1. Name your new Worker directory by specifying where you want to create your application.
2. Select `"Hello World" script` as the type of application you want to create.
3. Answer `no` to using TypeScript.
4. Answer `no` to using Git.
5. Answer `no` to deploying your Worker.

This will create a new directory, which will include both a `src/worker.js` file to write your code, and a [`wrangler.toml`](/workers/wrangler/configuration/) configuration file. After you create your Worker, create a class to define a Durable Object.

> ## 3. Configure Durable Object bindings

Reviewer 2: This should be step 5.

> Configure Durable Object bindings in your Worker project's `wrangler.toml` by providing a binding name (for this guide, use `EXAMPLE_CLASS`) and the class name (`DurableObjectExample`).

Reviewer 2: ```suggestion
[Bindings](/workers/configuration/bindings/) allow your Workers to interact with resources on the Cloudflare developer platform. Configure Durable Object bindings in your Worker project's `wrangler.toml` by providing a binding name (for this guide, use `EXAMPLE_CLASS`) and the class name (`DurableObjectExample`).
```

```toml
---
filename: wrangler.toml
---

> [durable_objects]
> bindings = [
>   { name = "EXAMPLE_CLASS", class_name = "DurableObjectExample" } # Binding to our DurableObjectExample class

Reviewer 2: ```suggestion
[[durable_objects.bindings]]
name = "EXAMPLE_CLASS"
class_name = "DurableObjectExample"
```

]
```

> The `[durable_objects]` section has one subsection called `bindings`, which is an array of tables. 
> 
> Each table contains the following fields:

Reviewer 2: ```suggestion
The `[[durable_objects.bindings]]` section contains the following fields:
```

  - `name` - Required. The binding name to use within your Worker.
  - `class_name` - Required. The class name you wish to bind to.
  - `script_name` - Optional. Defaults to the current [environment's](/workers/wrangler/environments/) script.

> ## 4. Configure Durable Object classes with migrations

Reviewer 2: This should be step 6.

A migration is a mapping process from a class name to a runtime state. You perform a migration when creating a new Durable Object class, renaming, deleting and transferring an existing Durable Object class. 

Migrations are performed through the `[[migrations]]` configurations key in your `wrangler.toml` file.  

> To configure a Durable Object migration in `wrangler.toml` file:

Reviewer 2: ```suggestion
To configure a Durable Object migration in your Worker's `wrangler.toml` file:
```

```toml
---
filename: wrangler.toml
---
[[migrations]]
tag = "v1" # Should be unique for each entry
new_classes = ["DurableObjectExample"] # Array of new classes
```
Refer to [Durable Objects migrations](/durable-objects/learning/durable-objects-migrations/) to learn more about the migration process.

>  ## 5. Write a class to define a Durable Object

Reviewer 2: This should be step 3.

Before you create and access a Durable Object, you must define its behavior by exporting an ordinary JavaScript class. 

{{<Aside type="note">}}
If you do not use JavaScript, you will need a [shim](https://developer.mozilla.org/en-US/docs/Glossary/Shim) to translate your class definition to a JavaScript class.
{{</Aside>}}

> Your class `DurableObjectExample` will have a constructor with two parameters. The first parameter, `state`, passed to the class constructor contains state specific to the Durable Object, including methods for accessing storage. The second parameter, `env`, contains any bindings you have associated with the Worker when you uploaded it. 

Reviewer 2: ```suggestion
Your `DurableObjectExample` class will have a constructor with two parameters. The first parameter, `state`, passed to the class constructor contains state specific to the Durable Object, including methods for accessing storage. The second parameter, `env`, contains any bindings you have associated with the Worker when you uploaded it. 
```

> Note that this means bindings are no longer global variables. For example, if you had a secret binding `MY_SECRET`, you must access it as `env.MY_SECRET`.

Reviewer 2: ```suggestion
Note that this means bindings are no longer global variables. For example, if you had a secret binding with a `binding` name of `MY_SECRET`, you must access it as `env.MY_SECRET`.
```

```js
---
filename: worker.js
---
export class DurableObjectExample {
  constructor(state, env) {}
}
```
Workers communicate with a Durable Object via the fetch API. Like a Worker, a Durable Object listens for incoming fetch events by registering an event handler. For a Durable Object, the fetch handler is defined as a method on the class.

> 

Reviewer 2: ```suggestion

Your `worker.js` file should now look like:

```

```js
---
filename: worker.js
---
export class DurableObjectExample {
  constructor(state, env) {}

  async fetch(request) {
    return new Response("Hello World");
  }
}
```

A Worker can pass information to a Durable Object via headers, the HTTP method, the Request body, or the Request URI.

{{<Aside type="note">}}

HTTP requests received by a Durable Object do not come directly from the Internet. HTTP requests come from other Workers and other Durable Objects. Durable Objects use HTTP for familiarity, but plan to introduce other protocols in the future.

{{</Aside>}}

> 

Reviewer 2: ```suggestion
```

> ## 6. Instantiate and communicate with a Durable Object

Reviewer 2: This should be step 4.

The `fetch()` handler allows you to instantiate and communicate to a Durable Object from a Worker. 

{{<Aside type="note">}}
Durable Objects do not receive requests directly from the Internet. Durable Objects receive requests from Workers or other Durable Objects. 
This is achieved by configuring a binding in the calling Worker for each Durable Object class that you would like it to be able to talk to. These bindings must be configured at upload time. Methods exposed by the binding can be used to communicate with particular Durable Object instances.
{{</Aside>}}

> To implement a `fetch()` handler, export a method named `fetch()` in an `export default {}` block:

Reviewer 2: ```suggestion
To implement a `fetch()` handler, export a method named `fetch()` in an `export default {}` block. Add the following to your `worker.js` file's existing class and fetch handler:
```

5. Forwarded the request to the Durable Object. `stub.fetch()` has the same signature as the global `fetch()` function, except that the request is always sent to the object, regardless of the request's URL.  The first time you send a request to a new object, the object will be created for us. If you do not store durable state in the object, it will automatically be deleted later (and recreated if you request it again). If you store durable state, then the object may be evicted from memory but its durable state will be kept  permanently.
6. Received an HTTP response back to the client with `return response`.

Refer to [Access a Durable Object from a Worker](/durable-objects/how-to/access-durable-object-from-a-worker/) to learn more about communicating to a Durable Object.

## 7. Upload a Durable Object Worker

To upload Workers that implement or bind to Durable Objects, use [Wrangler](/workers/wrangler/), the Workers CLI. 

Open your terminal and run the following command to upload a Durable Object Worker:

```sh
$ wrangler dev
```

## 8. Deploy your Durable Object Worker

> Use [Wrangler](/workers/wrangler/), the Workers CLI to deploy your Durable Object.
> 
> Open your terminal and run the following command to deploy a Durable Object Worker:

Reviewer 2: ```suggestion
To deploy your Durable Object Worker:
```

```sh
$ wrangler deploy
```

> Once deployed, you should be able to see your newly created Durable Object on the [Cloudflare dashboard](https://dash.cloudflare.com/), **Workers & Pages** > **Overview**.

Reviewer 2: ```suggestion
Once deployed, you should be able to see your newly created Durable Object Worker on the [Cloudflare dashboard](https://dash.cloudflare.com/), **Workers & Pages** > **Overview**.

Preview your Durable Object Worker at `<YOUR_WORKER>.<YOUR_SUBDOMAIN>.workers.dev`.
```

By finishing this tutorial, you have successfully created, tested and deployed a Durable Object.

> 

Reviewer 2: ```suggestion
```

### Related resources

> 

Reviewer 2: ```suggestion
```

{{<Aside type="note">}}

HTTP requests received by a Durable Object do not come directly from the Internet. They come from other Worker code – possibly other Durable Objects, or Workers. Durable Objects use HTTP for familiarity, but plan to introduce other protocols in the future.

{{</Aside>}}

## 4. Configure Durable Object bindings

Configure Durable Object bindings in the `wrangler.toml` by providing the class name and script name whose objects you wish to access using the binding. The script name can be omitted when creating a binding for a class that is defined in the same Worker as the binding.

```toml
---
filename: wrangler.toml
---
[durable_objects]
bindings = [
  { name = "EXAMPLE_CLASS", class_name = "DurableObjectExample" } # Binding to our DurableObjectExample class
]
```

The `[durable_objects]` section has one subsection called `bindings`, which is an array of tables. 

Each table contains the following fields:

  - `name` - Required. The binding name to use within your Worker.
  - `class_name` - Required. The class name you wish to bind to.
  - `script_name` - Optional. Defaults to the current [environment's](/workers/wrangler/environments/) script.

> If you are using Wrangler [environments](/workers/wrangler/environments/), you must specify any Durable Object bindings you wish to use on a per-environment basis. 

Reviewer 2: This needs to be another page in the docs. Environments information is not necessary for a get started. Everything from here to `## 5.` could and should be moved.

Durable Object bindings are not inherited. For example, you can define an environment named `staging` as below:

```toml
---
filename: wrangler.toml
---
[env.staging]
durable_objects.bindings = [
  {name = "EXAMPLE_CLASS", class_name = "DurableObjectExample"}
]
```

Because Wrangler appends the [environment name](/workers/wrangler/environments/) to the top-level name when publishing, for a Worker named `worker-name` the above example is equivalent to:

```toml
---
filename: wrangler.toml
---
[env.staging]
durable_objects.bindings = [
  {name = "EXAMPLE_CLASS", class_name = "DurableObjectExample", script_name = "worker-name-staging"}
]
```

> `"EXAMPLE_CLASS"` in the staging environment is bound to a different script name compared to the top-level `"EXAMPLE_CLASS"` binding, and will therefore access different objects with different persistent storage. 

Reviewer 2: This is also refering to Workers as 'scripts'  which we do not do anymore. This will need a refresh.

If you want an environment-specific binding that accesses the same objects as the top-level binding, specify the top-level script name explicitly:

```toml
---
filename: wrangler.toml
---
[env.another]
durable_objects.bindings = [
  {name = "EXAMPLE_CLASS", class_name = "DurableObjectExample", script_name = "worker-name"}
]
```

## 5. Configure Durable Object classes with migrations

> Migrations are performed through the `[[migrations]]` configurations key in your `wrangler.toml` file.  

Reviewer 2: What are migrations?

3. Derived an object ID from the URL path. `EXAMPLE_CLASS.idFromName()` always returns the same ID when given the same string as input (and called on the same class), but never the same ID for two different strings (or for different classes). In this case, you are creating a new object for each unique path. 
4. Constructed the stub for the Durable Object using the ID. A stub is a client object used to send messages to the Durable Object.
5. Forwarded the request to the Durable Object. `stub.fetch()` has the same signature as the global `fetch()` function, except that the request is always sent to the object, regardless of the request's URL.  The first time you send a request to a new object, the object will be created for us. If you do not store durable state in the object, it will automatically be deleted later (and recreated if you request it again). If you store durable state, then the object may be evicted from memory but its durable state will be kept  permanently.
6. Received an HTTP response back to the client with `return response`.

> Refer to [Access a Durable Object from a Worker](/durable-objects/how-to/access-durable-object-from-a-worker/) to  learn more about communicating to a Durable Object.

Reviewer 2: ```suggestion
Refer to [Access a Durable Object from a Worker](/durable-objects/how-to/access-durable-object-from-a-worker/) to learn more about communicating to a Durable Object.
```

## 7. Upload a Durable Object Worker

> {{<Aside type="warning" header="Custom Wrangler installation instructions">}}
> 
> You must use [Wrangler latest version](/workers/wrangler/install-and-update/) to manage Durable Objects.
> 
> {{</Aside>}}

Reviewer 2: ```suggestion
```
The user has already installed Wrangler in step 1. If we want to put version requirements, let's put them at the top of the guide.

> To upload Workers that implement or bind to Durable Objects, use [Wrangler](/workers/wrangler/), the Workers CLI. 

Reviewer 2: Are we uploading as in running locally? Or uploading as in deploying? This part is unclear.

Open your terminal and run the following command to upload a Durable Object Worker:

```sh
$ wrangler dev
```

> ### Specify the main module

Reviewer 2: This seems like instructions that need to be at the top with the rest of the wrangler.toml file modification.

Workers that use ES Modules syntax must have a main module specified from which all Durable Objects and event handlers are exported. The file that should be treated as the main module is configured using the `main = "src/worker.js"` section of `wrangler.toml`. 

Refer to [Custom builds documentation](/workers/wrangler/custom-builds/) for more details.

> ## 8. Test your Durable Objects project

Reviewer 2: Where does `wrangler deploy` fit into this flow?

A binding is defined in the `wrangler.toml` file of your Worker project’s directory.

{{<Aside type="note" header="What is a binding?">}}

A binding is a how your Worker interacts with external resources such as [KV Namespaces](/workers/runtime-apis/kv/) or Durable Objects. A binding is a runtime variable that the Workers runtime provides to your code.

You can declare a variable name in your `wrangler.toml` file that will be bound to these resources at runtime, and interact with them through this variable. Every binding’s variable name and behavior is determined by you when deploying the Worker. Refer to the [Environment Variables](/workers/platform/environment-variables/) documentation for more information.

{{</Aside>}}

A Worker talks to a Durable Object through a stub object. The class binding's `get()` method returns a stub to the particular Durable Object instance, and the stub's `fetch()` method sends HTTP [Requests](/workers/runtime-apis/request/) to the instance.

> The fetch handler in the example below implements the Worker that talks to the Durable Object. Note that the fetch handler is written using a new kind of Workers syntax based on ES Modules. This syntax is required for scripts that export Durable Objects classes, but is not required for scripts that make calls to Durable Objects. However, Workers written in the modules syntax (including Durable Objects) cannot share a script with Workers written in the Service Worker syntax.

Reviewer 3: Do we still want to mention this? I feel like it'd be better to just say you need to use ES Modules syntax for scripts that implement DOs, and then as an "Aside" we can mention that calling workers (in other scripts) can be written in Service Worker syntax.

Point being it doesn't seem as important to make this clarification as it might have been earlier during DO open beta (or whenever we started using ES Module syntax). CC @elithrar

We recommend following this approach of implementing Durable Objects and a corresponding fetch handler in the same script (written in the modules format) for convenience, but it is not required.

> ES Modules differ from regular JavaScript files in that they have imports and exports. When [writing the class that defines a Durable Object](/durable-objects/get-started/#2-write-a-class-that-defines-a-durable-object), you wrote `export class DurableObjectExample` when creating your class. To implement a fetch handler, export a method named `fetch()` in an `export default {}` block.

Reviewer 3: It feels a bit awkward to have this be so directed to the reader (in particular, saying _"you wrote"_ and _"your class"_). I've looked at the "learning DOs" page a number of times without actually writing anything so reading that would slightly interrupt me imo. Can we change it to say _"we wrote"_ and _"our class"_?

Reviewer 3: SGTM!
